CIPs/CIP-71/CIP-71.md

```diff
@@ -0,0 +1,184 @@
+
+```typescript
+function loadDocuments(id: DocID, paths: string[]): Promise<Record<DocID, Doctype>>
+```
+
+For example if *doc A* links to *doc B* on the `coolLink` property in its content `ceramic.loadDocuments(<DocID(doc A)>, ['/coolLink'])` would return:
+
+```typescript
+{
+  <DocID(doc A)>: <content(doc A)>,
+  <DocID(doc B)>: <content(doc B)>
+}
+```
+
+If some of the given `paths` are not present in the document the method should just return the documents that where actually present. This will allow developers to optimistically query for multiple linked documents and only get the already created docs.
+
+### MultiQuery on http-daemon
+
+The http daemon should be extended with a `states` method. This method allows for multiple Multiqueries to be done at once. The reason to add the ability to make multiple queries at once is that the http-client often wants to maintain the state of multiple documents at once in the background, so this method can be used to reduce the number of requests made.
+
+**Endpoint:** `POST /api/v0/states`
+
+**Body:**
+
+The body should conform to the `StatesQuery` interface.
+
+```typescript
+interface MultiQuery {
+  docid: string
+  paths?: Array<string>
+}
+
+interface StatesQuery {
+  queries: Array<MultiQuery>
+}
+```
+
+**Response:**
+
+The body should conform to the `Response` interface.
+
+```typescript
+interface ResponseData {
+  [docid: string]: DocState
+}
+
+interface Response {
+  status: 'success'
+  data: ResponseData
+}
+```
+
+### Pubsub Queries
+
+Replace the current message types with the ones defined below. 
+
+```typescript
+interface Update {
+  typ: 0
+  doc: string
+  tip: string
+  anchorService?: string
+}
+
+interface Query {
+  typ: 1
+  id: string
+  doc: string
+  paths?: Array<string>
+}
+
+interface TipMap {
+  [docid: string]: string
+}
+
+interface Response {
+  typ: 2
+  id: string
+  tips: TipMap
+}
+```
+
+#### Update
+
+The new update message merges the previous `UPDATE` and `ANCHOR_META` message types by adding an optional `anchorService` property to the mesasge. This property will be set if an anchor has been requested but not yet returned. When an anchor do get returned a new update message will be sent with the CID of the *anchorRecord* as the `tip`, and `anchorSerivce` not set.
+
+**Properties:**
+
+* `typ` - the message type, always `0`
+* `doc` - the DocID that is being updated
+* `tip` - the CID of the new Tip of the document
+* `anchorService` - the url of the anchor service that was used to request an anchor
+
+#### Query
+
+The query message is a replacement of the old `REQUEST` message type. The most important aspect is the new format for the `id` and the new property `paths`.
+
+**Properties:**
+
+* `typ` - the message type, always `1`
+* `id` - unique identifier of the query, multihash of the query object without the `id` property
```

Comment:
should we specify here how this multihash is being built? do we need to canonicalize the query for that, etc?

Reply:
good point.